### ISSUE TITLE
Standalone Quorum DDS

### ISSUE THREAD

**Issue by ChumpChief** [microsoft/FluidFramework#9049]
NOTE: We expect a standalone Quorum DDS may be generally useful for a variety of scenarios.  We are currently investigating using it as a mechanism to coordinate safely ending collaboration, as described below.

# Scenario: Safely ending collaboration

Suppose a customer wants to export data from a container with high confidence that no connected client has outstanding unsaved ops.  This is difficult to achieve with the tools we provide today.  The exporting client needs to signal to other clients in some way that they should "save and quit", and then those clients need to signal back that they have completed "save and quit".

- Using a CRC isn't sufficient, since it only indicates that the signal was ack'd by the server, not by the other connected clients.
- A convention could be made using maps, for clients to explicitly sign off that they had completed save/quit, but this puts a reasonable burden on customers to implement.  There may also be complications to consider in op replay scenarios.
- A Quorum-like data structure (using MSN as a signal that connected clients have seen the request) seems somewhat attractive as a "more-automatic" approach than using maps.  There are potentially complications to consider if the other clients need to complete async work before saving and quitting.

Whatever approach we choose, we should have a clear best practice we can promote to our customers who want to achieve this scenario.

# Work breakdown

Update:  As I've investigated this further, I'm realizing some shortcomings of the Quorum that are undesirable for the scenario.  Most notably is that it doesn't have particularly well-suited guarantees around simultaneous proposals from different clients on the same key.  Quorum does reach eventual consistency since the promotion to an "approved" proposal doesn't look at local un-ack'd state, BUT because of the last-write-wins resolution it becomes easy for the clients to clobber each others changes (which is never desirable in this scenario).

In this scenario I think a consensus-type resolution would be a better fit.  It's already no slower since Quorum already waits on the ack+MSN update anyway, and it eliminates the possibility of accidental clobber.  Actually, what we want is probably even stronger than consensus-type resolution, because we'll want to discard subsequent writes all the way up until the proposal is approved (not just sequenced).

Extracting the Quorum proposals functionality for reuse
- [x] #9058 - #9092
- [x] #9059 - #9131
- [x] #9137 - #9224
- [x] #9060 - #9661
- [x] #10027 - #10028
- [x] #10061
- [ ] Refine Quorum API surface and incorporate feedback
- [x] Consider explicit ack approach rather than MSN-based approach
- [x] Consider "request for delay" API surface (if not implicitly covered by explicit ack approach)
- [ ] Consider rejection API surface
- [ ] Promote to full (non-experimental) package

Optional followup
- [ ] Consider usage in `Container` for the built-in `Quorum`, or alternate approaches to code proposal storage.

**Comment by vladsud** [comment]
Given that clients can disconnect at any moment, I'm not sure I'd try to have such capability. I'd rather assume that whatever capability app provides is equivalent to Save-As, it's a state of the document at some point, and doc itself just keeps moving forward, independently from agent doing Save-As.

Also, client that does export may die or hang. Does it block all clients on doc forever?

Unless it's equivalent to "this file has transitioned to read-only state" functionality, I'm not sure I can see how it can be accomplished.

**Comment by ChumpChief** [comment]
Just for context, this is generally a follow-on to the external data sync investigation I did late last year.

Being robust to client death/hang is part of the scope of this.  With this mechanism, the collaboration comes to a halt so there's no moving forward - it's not intended to be "save as" functionality.  Rather, it's part of a strategy to enable apps to safely enact non-compat schema changes on existing documents ("landing the planes" as we've discussed previously).  I'll be putting some material together on the bigger picture for this.

**Comment by vladsud** [comment]
I know it's hard to do in short period of time, but I want to point out that we need more general mechanism that does not involve stopping collaboration. The trivial example - we likely not be able to avoid running 3-way merge when coming from offline as a way to reduce impact of losing user intent through ops, and as a way of detecting conflicts and presenting user choice to resolve them. The result of such operation would be new document state, sort of revision, that is impossible to decompose into set of ops that make sense (from app perspective, as there is no way to define "right" order for such ops, at least if they are at granularity of current ops). So, we will need to build capability to commit such new rev and reason about how it interacts with ops that are racing with such update.

In addition to that, creating such mechanism would require deep runtime involvement, including changes in PendingStateManager (who today submits any pending ops on reconnect right away), which is not great (I'd rather prefer it to be outside of core runtime, as a temp solution until we solve above problem).

The way this EPIC is described right now - it's about export, and export does not need such capability.
Import is different, if that's the scenario, we should update task.
But as in other places, I've pointed out that the problem of overwriting file exists today no matter what and lies outside of just client capabilities, as data center can restore to previous state, can do fail over to previous replicate / different geo, etc. So there has to be mechanism to detect that file overwritten in some way. Currently for SPO it's epoch, though there is some work happening (or rather frozen) to move to op check sums (though I'm not convinced it will fully remove need for epoch, it likely will compliment it). And once we have that mechanism for FRS, we should consider if it can be leveraged here, where clients doing conversion are Ok with possibly losing couple seconds of latest ops but can overwrite a file and force all the other clients to fail and reload.

Basically, whether there is some consensus here or not, there is data loss in either case, as clients who just disconnected or in process of converting to "write" connection have data loss. So, I'd rather optimize for simplicity, as we can't solve this problem without proper 3-way merge in place.

**Comment by ChumpChief** [comment]
Remaining work is tracked in ADO, experimental version is available as PactMap.
